feat(table): Add order functionality to table.

peter-affenzeller opened the pull request:
### <strong>Pull Request</strong>

#### Type of PR

<!-- Bugfix (non-breaking change which fixes an issue) -->
Feature (non-breaking change which adds functionality)
<!-- Breaking change (fix or change that would cause existing functionality to not work as expected) -->
<!-- Documentation update (changes to documentation) -->
<!-- Other (if none of the above apply) -->

#### Checklist

- [x] I have read the CONTRIBUTING doc and I follow the PR guidelines
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

peter-affenzeller commented on libs/barista-components/table/src/order/order.ts:
```diff
@@ -0,0 +1,111 @@
+export class DtOrder<T> extends _DtOrderMixinBase
+  implements CanDisable, AfterViewInit, OnChanges, OnDestroy {
+  /** Event that is emitted whenever the input value of the order column changes. */
+  @Output('dtOrderChange') readonly orderChange = new EventEmitter<
```
yeah, does fire in both cases, I'll change the comment to make that clearer

peter-affenzeller commented on libs/barista-components/table/src/table-module.ts:
```diff
@@ -91,13 +98,15 @@ const EXPORTED_DECLARATIONS = [
     DtIndicatorModule,
     DtFormattersModule,
     DtEmptyStateModule,
+    ReactiveFormsModule,
```
I use a form control for validating the input field of the order cell

lukasholzer replied:
the normal FormsModule should be enough for your use case

peter-affenzeller replied:
no, I can't bind to formControl if I switch to the FormsModule, or am I missing something?

ffriedl89 replied:
It's perfectly fine to use the ReactiveFormsModule in my opinion, since we have @angular/forms as a peerDep already - so if you want to use the formControl internally that's ok for me :)

peter-affenzeller commented:
@ffriedl89 UX pushed a new drag handle icon, I'll change that as soon as it is available.
I implemented the suggested changes, the discussed indicator animation for moved rows after an input change is also included.

thomaspink commented on libs/barista-components/table/src/order/order-cell.ts:
```diff
@@ -0,0 +1,158 @@
+  /** Subject used for unsubscribing */
+  private _destroy$ = new Subject<void>();
+
+  @ViewChild('orderInput', { static: true }) _orderInput: ElementRef<
```
NIT: If you format it that way you do not need to linebreak the generic
```
@ViewChild('orderInput', { static: true }) 
_orderInput: ElementRef<HTMLInputElement>;
```

thomaspink commented on libs/barista-components/table/src/order/order-cell.ts:
```diff
@@ -0,0 +1,158 @@
+  ]);
+
+  constructor(
+    public _columnDef: DtColumnDef,
+    public _changeDetectorRef: ChangeDetectorRef,
```
Does not need to be public
```suggestion
    columnDef: DtColumnDef,
    changeDetectorRef: ChangeDetectorRef,
```

thomaspink commented on libs/barista-components/table/src/table-order-data-source.ts:
```diff
@@ -0,0 +1,119 @@
+  private _renderChangesSubscription = Subscription.EMPTY;
+
+  /** Array of data that should be rendered by the table, where each object represents one row. */
+  get data(): T[] {
```
Why do we need to reimplement this if we extend from the datasource which already holds the exact same logic?

peter-affenzeller replied:
changed the order data source to extend from the table data source

peter-affenzeller replied:
@thomaspink I actually changed this back to extend from the cdk data source, else the user would still be able to call things related to sorting, which is what we wanted to avoid.

tomheller replied:
CDK datasource does not implement data or the constructor, so I don't see a problem with this implementation:

CDK table datasource export:
https://github.com/angular/components/blob/14a51ef07e8ec4db4696c2dd4b4fcbb295dfcb24/src/cdk/table/public-api.ts#L18-L19

DataSource in the cdk collections:
https://github.com/angular/components/blob/14a51ef07e8ec4db4696c2dd4b4fcbb295dfcb24/src/cdk/collections/data-source.ts#L12-L31

thomaspink commented on libs/barista-components/table/src/table-order-data-source.ts:
```diff
@@ -0,0 +1,119 @@
+
+export class DtTableOrderDataSource<T> extends DataSource<T> {
+  /** Stream that emits when a new data array is set on the data source. */
+  private _data: BehaviorSubject<T[]>;
```
Why do we need to reimplement this if we extend from the datasource which already holds the exact same property?

thomaspink commented on libs/barista-components/table/src/table-order-data-source.ts:
```diff
@@ -0,0 +1,119 @@
+  }
+  private _order: DtOrder<T> | null;
+
+  constructor(initialData: T[] = []) {
```
Why do we need to reimplement this if we extend from the datasource which already holds the exact same logic?

thomaspink commented on libs/barista-components/table/src/table-order-data-source.ts:
```diff
@@ -0,0 +1,119 @@
+   * Used by the DtTable. Called when it connects to the data source.
+   */
+  connect(_table: DtTable<T>): Observable<T[]> {
+    return this._data;
```
Why isn't the ordered table paginated (like in the table data source; see filteredData)?

peter-affenzeller replied:
that was something that also came up in the review and @ursula-wieshofer suggested that pagination in combination with drag&drop usually leads to UX issues and should be avoided.

lukasholzer commented:
![dependencies](https://user-images.githubusercontent.com/11156362/80955268-4b6f8c80-8dff-11ea-801f-306662925171.png)

tomheller commented on libs/barista-components/table/src/order/order-cell.scss:
```diff
@@ -0,0 +1,121 @@
+  align-items: center;
+}
+
+::ng-deep .dt-simple-order-column-icon {
```
can we prefix a host here or do we need to escape the encapsulation here because of the dragging.

tomheller commented on libs/barista-components/table/src/order/order-cell.scss:
```diff
@@ -0,0 +1,121 @@
+  height: 1.5rem !important;
+}
+
+::ng-deep .dt-table:not(.dt-order-disabled) .dt-simple-order-column-icon {
```
can we prefix a host here or do we need to escape the encapsulation here because of the dragging.

tomheller commented on libs/barista-components/table/src/order/order-cell.scss:
```diff
@@ -0,0 +1,121 @@
+  cursor: pointer;
+}
+
+::ng-deep .dt-simple-order-column-icon svg {
```
can we prefix a host here or do we need to escape the encapsulation here because of the dragging.

tomheller commented on libs/barista-components/table/src/order/order-cell.scss:
```diff
@@ -0,0 +1,121 @@
+
+::ng-deep .dt-simple-order-column-icon {
+  margin-right: 4px;
+  width: auto !important;
+  height: 1.5rem !important;
```
Can we do without the important here? maybe double up on the class selector

tomheller commented on libs/barista-components/table/src/order/order-cell.scss:
```diff
@@ -0,0 +1,121 @@
+  fill: $turquoise-600;
+}
+
+::ng-deep .dt-table.dt-order-disabled .dt-simple-order-column-icon svg {
```
can we prefix a host here or do we need to escape the encapsulation here because of the dragging.

tomheller commented on libs/barista-components/table/src/order/order-cell.scss:
```diff
@@ -0,0 +1,121 @@
+}
+
+.dt-simple-order-input-invalid {
+  background: $red-200 !important;
+  border: 1px solid $error-color !important;
```
Can we remove the importants?

tomheller commented on libs/barista-components/table/src/row.ts:
```diff
@@ -29,9 +29,21 @@ import {
+  keyframes,
+} from '@angular/animations';
+
+// prettier-ignore
```
```suggestion
```

tomheller commented on libs/barista-components/table/src/row.ts:
```diff
@@ -54,17 +66,40 @@ export class DtRowDef<T> extends CdkRowDef<T> {}
    */
   static _mostRecentRow: DtRow | null = null;
 
+  _orderChangedIndicatorAnimationState: 'void' | 'changed' = 'void';
```
missing jsDoc

peter-affenzeller commented on apps/components-e2e/src/components/table-order/table.e2e.ts:
```diff
@@ -0,0 +1,60 @@
+    .expect(dataCells.nth(0).textContent)
+    .eql('I')
+    .expect(dataCells.nth(1).textContent)
+    .eql('II');
```
I'll add a couple more for the input fields and whatever I can think of that makes sense, sure, forgot about that after fighting some other issues...

peter-affenzeller commented on libs/barista-components/table/src/order/order-cell.scss:
```diff
@@ -0,0 +1,121 @@
+  }
+}
+
+::ng-deep {
```
in this case it is needed to escape encapsulation because of the dragging, yes.

peter-affenzeller commented:
@tomheller I changed the styling to get rid of the importants and add prefixes if possible. also included the dt-main-font mixin to fix the font-size of the drag preview (changed slightly in the barista examples)
added the examples in the readme and included more e2e tests.

@ffriedl89 since you can enter a very high number in the input to move a row to the last position, I changed the order-directive slightly to clamp the value to the number of rows when emitting the order change event. else it would emit e.g. previousIndex 0, currentIndex 999 even though there are only 10 rows in the table.

the build succeeds now, thx @lukasholzer for helping!
